contractcourt/channel_arbitrator.go

```diff
@@ -1108,31 +1110,77 @@ func (c ChainActionMap) Merge(actions ChainActionMap) {
-// last minute as we want to ensure that when we *need* (HTLC is timed out) to
-// sweep, the commitment is already confirmed.
-func (c *ChannelArbitrator) shouldGoOnChain(htlcExpiry, broadcastDelta,
-	currentHeight uint32) bool {
+// shouldGoOnChain returns a bool indicating whether we should go on chain
+// to claim a htlc. It factors in two considerations:
+// - confirmation delta: the number of blocks that we estimate that we will
+// need to confirm the commitment transaction. If the htlc is going to time
+// out within this delta from the current block height, we consider going on
+// chain. If not, the htlc is not (yet) at risk of timing out before the
+// commitment is confirmed so we do not need to go on chain.
+// - htlc sweep cost: if we are considering going on chain because the htlc
+// timeout is within the broadcast delta, we check whether it would be
+// economical to do so. If the htlc amount less than our current estimate
+// of what it would cost to sweep, we do not recommend going on chain. Note
+// that current fees are not representative of future fees, and we may not
+// have to sweep the htlc (if the remote peer sweeps the success path for
+// outgoing htlcs, or the timeout path for incoming htlcs), so this assessment
+// is an approximate one.
+func (c *ChannelArbitrator) shouldGoOnChain(htlc channeldb.HTLC,
+	broadcastDelta, currentHeight uint32,
+	feeEstimator func(blocks uint32) (
+		chainfee.SatPerKWeight, error)) (bool, error) {
 
 	// We'll calculate the broadcast cut off for this HTLC. This is the
 	// height that (based on our current fee estimation) we should
 	// broadcast in order to ensure the commitment transaction is confirmed
 	// before the HTLC fully expires.
-	broadcastCutOff := htlcExpiry - broadcastDelta
-
-	log.Tracef("ChannelArbitrator(%v): examining outgoing contract: "+
-		"expiry=%v, cutoff=%v, height=%v", c.cfg.ChanPoint, htlcExpiry,
-		broadcastCutOff, currentHeight)
+	broadcastCutOff := htlc.RefundTimeout - broadcastDelta
 
 	// TODO(roasbeef): take into account default HTLC delta, don't need to
 	// broadcast immediately
 	//  * can then batch with SINGLE | ANYONECANPAY
 
-	// We should on-chain for this HTLC, iff we're within out broadcast
-	// cutoff window.
-	return currentHeight >= broadcastCutOff
+	// If we are not at the broadcast cutoff height yet, do not worry
+	// about going to chain.
+	if currentHeight < broadcastCutOff {
+		return false, nil
+	}
+
+	// If we are on or beyond the broadcast cutoff height, which means we
+	// are at risk of the commitment not confirming before the htlc times
+	// out. It does not make sense to go on chain with a htlc which will
+	// cost more to sweep than it is worth, so we estimate what sweeping
+	// will cost to determine whether we should go on chain.
+
+	// Get the fee rate estimate (sat/kweight) that we estimate would be
+	// required to get this htlc on chain within the broadcast cutoff.
+	feeRateEstimate, err := feeEstimator(broadcastCutOff)
```

Comment:
I think we should use the fee that we are actually going to use if we'd go to chain now. Currently that is a fixed conf target that ignores how close we are to expiry.

Reply:
Also, if we force close, the htlcs need to be claimed first with the pre-signed second level transaction for which the fee is already set.